---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%",
  cache = TRUE,
  warning = FALSE,
  message = FALSE,
  cache.lazy = FALSE
)
```

# GeneSPLS

<!-- badges: start -->
<!-- badges: end -->

The goal of GeneSPLS is to provide an implementation of Sparse Partial Least Squares (SPLS) regression tailored to genome-wide association studies (GWAS) data analyses. SPLS is particularly useful for high-dimensional datasets where the number of predictors (e.g., SNPs) far exceeds the number of individuals since SPLS induces sparsity in the set of selected model predictors. Also, GeneSPLS is implemented in Rcpp, which allows for faster computation compared to the original R implementation (see `spls` package), which is especially beneficial for large datasets. Finally, the underlying algorithm implemented here is Wide Kernel Partial Least Squares (PLS) regression, which is a variant of PLS (and by extension SPLS) that is computationally efficient for high-dimensional data.

## Installation

You can install the development version of GeneSPLS from [GitHub](https://github.com/) with:

``` r
# install.packages("pak")
pak::pak("syl051088/GeneSPLS")
```
## Main functions

- `spls_cpp()`: The main function for fitting SPLS regression models using Rcpp.
- `cv_spls_cpp()`: The main function for tuning SPLS regression model parameters using cross-validation.

## Example

This is a basic example which simulates genotype data with a univariate outcome, tunes model parameters using cross-validation, and fits an SPLS regression model with the selected parameters:

```{r example}
library(GeneSPLS)
library(tidyverse)
set.seed(815)

source(here::here("R", "helper.R"))

### SIMULATE DATA

# Simulating high-dimensional predictor matrix (p >> n)
n <- 200   # Number of individuals
p <- 5000  # Number of SNPs

# Simulated SNP genotype data (coded as 0, 1, or 2)
X <- matrix(sample(0:2, n * p, replace = TRUE), nrow = n, ncol = p)
colnames(X) <- paste0("SNP_", 1:p)

# Simulated phenotype (continuous trait)
Y <- X[, sample(1:p, 600)] %*% runif(600, -1, 1) + rnorm(n)


### PARAMETER CROSS-VALIDATION

# Define candidate parameters
eta_candidates <- seq(0.1, 0.9, by = 0.1)
K_candidates <- 1:10

# Perform cross-validation
cv_result_cpp <- cv_spls_cpp(
  X,
  Y,
  fold = 5,
  eta = eta_candidates,
  K = K_candidates,
  kappa = 0.5,
  select = "pls2",
  scale_x = TRUE,
  scale_y = FALSE,
  eps = 1e-4,
  maxstep = 100,
  trace = FALSE)

best_eta <- cv_result_cpp$eta.opt # Optimal sparsity parameter
best_K <- cv_result_cpp$K.opt # Optimal latent components


### FIT SPLS MODEL

spls.model <- spls_cpp(
  X,
  Y,
  K = best_K,
  eta = best_eta,
  kappa = 0.5,
  select = "pls2",
  fit = "widekernelpls",
  scale_x = TRUE,
  scale_y = FALSE,
  eps = 1e-4,
  maxstep = 100,
  trace = FALSE)

# Plot non-zero estimated coefficients
# Note: fit simple model for plot clarity
sens_plot(
  X,
  Y,
  eta = c(0.85),
  K = 1)
```


## File Structure

```text
GeneSPLS
├── DESCRIPTION
├── LICENSE
├── NAMESPACE
├── R
│   ├── RcppExports.R
│   └── helper.R
├── README.Rmd
├── README.md
├── man
│   ├── correctp.Rd
│   ├── cv_spls_cpp.Rd
│   ├── figures
│   ├── spls_cpp.Rd
│   ├── spls_dv.Rd
│   ├── ust.Rd
│   └── widekernelpls_fit.Rd
├── src
│   ├── GeneSPLS.dll
│   ├── GeneSPLS.so
│   ├── Makevars
│   ├── Makevars.win
│   ├── RcppExports.cpp
│   ├── RcppExports.o
│   ├── correctp.cpp
│   ├── correctp.h
│   ├── correctp.o
│   ├── cv.spls.cpp
│   ├── cv.spls.h
│   ├── cv.spls.o
│   ├── pls.cpp
│   ├── pls.h
│   ├── pls.o
│   ├── spls.cpp
│   ├── spls.h
│   ├── spls.o
│   ├── spls_dv.cpp
│   ├── spls_dv.h
    └── spls_dv.o
```